File: Introduction to Data Structures: Hash Table/5. Conclusion/2. Longest Substring Without Repeating Letters/LongestSubStringWithoutRepeatingCharacters1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        hashSet = set()
        l = 0
        r = 0
        maxLength = 0

        while r < len(s):

            if s[r] not in hashSet:

                hashSet.add(s[r])

                r += 1

                maxLength = max(maxLength, len(hashSet))

            else:

                hashSet.remove(s[l])

                l += 1

        return maxLength

File: Introduction to Data Structures: Hash Table/5. Conclusion/2. Longest Substring Without Repeating Letters/LongestSubStringWithoutRepeatingCharacters1.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        Set<Character> hashSet = new HashSet<>(); 
        
        int l = 0;
        
        int r = 0;
        
        int max = 0;
        
        while (r < s.length()) {
            
            if (!hashSet.contains(s.charAt(r))) {
                
                hashSet.add(s.charAt(r));
                
                max = Math.max(max, hashSet.size());
                
                r++; 
                
            } else {
                
                hashSet.remove(s.charAt(l));
                
                l++;
            }
        }
        
        return max;
    
    }
}